### PR TITLE
Update eks version

### LINKED ISSUE #22: Get EKS AMI's dynamically
In nodegroup controller we are currently hardcoding the value of AMI's. Should see if there is a way to get these values dynamically.

### LINKED ISSUE #85: Add Support for upgrading the EKS Version
Add update functionality for the Control Plane Version and the AMI running in the nodegroups for this.

### PR DESCRIPTION
**Pull request by Magizhchi**
*Issue #, if available:* resolves #85 resolves #22 

*Description of changes:* Add EKS Update for controlPlane and Nodegroups, Have also updated the nodegroups to pull the AMI from SSM instead of hardcoding the AMI's

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/controller/nodegroup/nodegroup_controller.go [hunk 1]
```diff
@@ -402,3 +408,9 @@ func shouldUpdate(crdParams []*cloudformation.Parameter, cfnParams []*cloudforma
 	}
 	return false
 }
+
+func getSSMParamKey(version string) string {
+	eksOptimizedAMIKey := "/aws/service/eks/optimized-ami/<EKS_VERSION>/amazon-linux-2/recommended/image_id"
```
Comment (anand99) [inline pkg/controller/nodegroup/nodegroup_controller.go]: nitpick: single use variable. i.e. we can directly use it in strings.Replace call without creating variable.

File: pkg/controller/nodegroup/nodegroup_controller.go [hunk 2]
```diff
@@ -341,31 +339,39 @@ func (r *ReconcileNodeGroup) updateNodeGroupStack(cfnSvc cloudformationiface.Clo
...
+			if ng.Spec.Instance.EBSVolumeSize != nil {
+				parameter = append(parameter, &cloudformation.Parameter{
+					ParameterKey:   aws.String("NodeVolumeSize"),
+					ParameterValue: aws.String(strconv.Itoa(*ng.Spec.Instance.EBSVolumeSize)),
+				})
+			}
+		}
 	}
 
-	if ngSpec.Instance.EBSVolumeSize != nil {
```
Comment (anand99) [inline pkg/controller/nodegroup/nodegroup_controller.go]: Are we removing ability to update based on volume size change? or it is just diff is messing up?

File: pkg/controller/controlplane/cfn_template.go [hunk 3]
```diff
@@ -5,6 +5,12 @@ var controlplaneCFNTemplate = `
 AWSTemplateFormatVersion: '2010-09-09'
 Description: 'Amazon EKS Networking + Control Plane [managed by aws-eks-cluster-controller]'
 
+Parameters:
+  EKSVersion:
+    Type: String
+    Description: Minimum size of Node Group ASG.
```
[the remark this answers is not in the document]
Comment (Magizhchi) [reply pkg/controller/controlplane/cfn_template.go]: Fixed

File: pkg/controller/controlplane/controlplane_controller.go [hunk 4]
```diff
@@ -282,3 +305,77 @@ func (r *ReconcileControlPlane) createControlPlaneStack(cfnSvc cloudformationifa
...
+	if len(output.Stacks) != 1 {
+		return nil, fmt.Errorf("Error while describing the stacks, got %d stacks for the name : %s", len(output.Stacks), stackName)
+	}
+
+	return output.Stacks[0].Parameters, nil
+}
+
+func shouldUpdate(crdParams []*cloudformation.Parameter, cfnParams []*cloudformation.Parameter) bool {
+	cfnParamMap := make(map[string]string)
+	for _, param := range cfnParams {
```
[the remark this answers is not in the document]
Comment (Magizhchi) [reply pkg/controller/controlplane/controlplane_controller.go]: The behaviour is slightly different from DeepEqual here, the equality is only one way. We only check if the CRD parameters are the same as the one's described in CFN, not the other way round. This is to account for default values, they will exist in CFN but not CRD

File: pkg/controller/nodegroup/nodegroup_controller.go (hunk 2 above)
Comment (Magizhchi) [reply pkg/controller/nodegroup/nodegroup_controller.go]: This looks good in my branch, diff is messing up, its removing and adding it few lines above

File: pkg/controller/controlplane/controlplane_controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (anand99) [reply pkg/controller/controlplane/controlplane_controller.go]: ah ok, got it.

**Pull request merged**